test/unittest/blas3/blas3_trsm_test.cpp

```diff
@@ -0,0 +1,95 @@
+/***************************************************************************
+ *
+ *  @license
+ *  Copyright (C) Codeplay Software Limited
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  For your convenience, a copy of the License has been included in this
+ *  repository.
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  SYCL-BLAS: BLAS implementation using SYCL
+ *
+ *  @filename blas3_gemm_test.cpp
+ *
+ **************************************************************************/
+
+#include "blas_test.hpp"
+
+template <typename scalar_t>
+using combination_t = std::tuple<int, int, char, char, char, char>;
+
+template <typename scalar_t>
+void run_test(const combination_t<scalar_t> combi) {
+  int m;
+  int n;
+  char transA;
+  char side;
+  char diag;
+  char triangle;
+  std::tie(m, n, transA, side, diag, triangle) = combi;
+
+  using data_t = utils::data_storage_t<scalar_t>;
+
+  scalar_t alpha = scalar_t{1};
+  const int lda = side == 'l' ? m : n;
+  const int ldb = m;
+  const int k = side == 'l' ? m : n;
+
+  const int sizeA = k * lda;
+  const int sizeB = n * ldb;
+
+  std::vector<scalar_t> A(sizeA);
+  std::vector<scalar_t> B(sizeB);
+
+  const scalar_t diagValue =
+      diag == 'u' ? scalar_t{1} : random_scalar(scalar_t{1}, scalar_t{10});
+
+  fill_triangle(A, k, k, lda, triangle, diagValue);
+  fill_random(B);
+
+  auto q = make_queue();
+  test_executor_t ex(q);
+  auto a_gpu = blas::make_sycl_iterator_buffer<scalar_t>(A, A.size());
+  auto b_gpu = blas::make_sycl_iterator_buffer<scalar_t>(B, B.size());
+
+  _trsm(ex, side, triangle, transA, diag, m, n, alpha, a_gpu, lda, b_gpu, ldb);
+
+  std::vector<data_t> X(B.size(), data_t{0});
+
+  auto x_gpu = blas::make_sycl_iterator_buffer<scalar_t>(X.size());
+  if (side == 'l') {
+    // Verification of the result. AX = B
+    _gemm(ex, transA, 'n', m, n, m, data_t{1}, a_gpu, lda, b_gpu, ldb,
```

Comment:
We should perhaps be using the reference blas TRSM to verify results here in line with other tests such as GEMM.

Reply:
The idea of using `_gemm` to verify the results is to simplify this test as it doesn't need to rely on a external library call to check if the result is correct. This is, in fact, just verifying that the output vector is in fact **a** solution for the system. If we change the internal algorithm later, this could potentially return a different, but valid, solution. Using _gemm just makes sure the result is a valid solution for the system, and _gemm is already tested.

Reply:
This is incorrect, depending on the type of TRSM the value for the upper or lower triangle can have an invalid value or can be padded with NAN. Your test should cover this case as well. If you use GEMM in this case you can get the wrong results.  Also, all other tests in the BLAS test against an external library. Please follow the same rule applied on other unit tests and add the nan tests and test against the external TRSM from the BLAS library that is already included in CMake.

Reply:
Is there an existing test that tests that uses NaN values for me to take as reference?

Reply:
> We should perhaps be using the reference blas TRSM to verify results here in line with other tests such as GEMM.

I've replaced the use of GEMM to verify the result with a call to system blas.

Reply:
Added test with NaN values in the input matrix